File: apps/server/gent_api/api/views/diff.py
"""
Commit diff endpoint.

Computes "what changed in a commit" entirely on the server by comparing the
commit's tree to its first parent's tree, then producing a unified line diff for
each changed text blob. This is the server-side counterpart of the algorithm the
web client previously had to run in the browser (it can now just call this).

Strategy (same shape as Git):
  1. `_diff_trees` walks both trees in lock-step, descending only into subtrees
     whose sha differs (identical subtrees are pruned), and emits one
     {added|removed|modified} record per changed file.
  2. For each changed text blob we load the old/new content and run a unified
     line diff via `difflib.SequenceMatcher.get_grouped_opcodes` (hunks + 3
     lines of context, exactly like `diff -u`).
"""
import difflib

from django.shortcuts import get_object_or_404
from rest_framework import status, permissions
from rest_framework.decorators import api_view, permission_classes
from rest_framework.response import Response
from drf_spectacular.utils import extend_schema
from drf_spectacular.types import OpenApiTypes

from api.models import Commit, Tree, Blob
from api.serializers import CommitSerializer
from api.utils import get_repository_or_404

# Above this many DP cells (len(old) * len(new)) we skip the inline line diff to
# keep the request bounded; the file is reported as changed with `too_large`.
MAX_DIFF_CELLS = 1_500_000


def _is_empty_sha(sha):
    """True for the all-zero placeholder sha given to empty trees/branches."""
    return not sha or set(sha) == {'0'}


def _split_lines(text):
    """Split into lines, normalising CRLF and dropping a single trailing newline."""
    if not text:
        return []
    text = text.replace('\r\n', '\n')
    if text.endswith('\n'):
        text = text[:-1]
    return text.split('\n')


def _load_entries(repository, tree_sha, cache):
    """Return a tree's entry list for a sha (cached), or [] if missing/empty."""
    if _is_empty_sha(tree_sha):
        return []
    if tree_sha in cache:
        return cache[tree_sha]
    tree = Tree.objects.filter(repository=repository, sha=tree_sha).first()
    entries = tree.entries if tree else []
    cache[tree_sha] = entries
    return entries


def _read_blob(repository, sha, cache):
    """Return (text, is_binary) for a blob sha. text is None when binary."""
    if sha in cache:
        return cache[sha]
    blob = Blob.objects.filter(repository=repository, sha=sha).first()
    result = ('', False)
    if blob is not None:
        content = blob.content
        if not content and blob.file_path:
            import base64  # local import: only on-disk blobs need it
            try:
                with open(blob.file_path, 'rb') as fh:
                    raw = fh.read()
                try:
                    content = raw.decode('utf-8')
                except UnicodeDecodeError:
                    result = (None, True)
                    cache[sha] = result
                    return result
            except (FileNotFoundError, OSError):
                content = ''
        result = (content or '', False)
    cache[sha] = result
    return result


def _collect_side(repository, entry, path, change_status, tree_cache, out):
    """Emit every blob under one side of an added/removed entry."""
    if entry['type'] != 'tree':
        record = {'path': path, 'status': change_status}
        record['new_sha' if change_status == 'added' else 'old_sha'] = entry['sha']
        out.append(record)
        return
    for sub in _load_entries(repository, entry['sha'], tree_cache):
        _collect_side(repository, sub, f"{path}/{sub['name']}", change_status, tree_cache, out)


def _diff_trees(repository, old_sha, new_sha, prefix, tree_cache, out):
    """Recursively diff two trees, pruning identical subtrees by sha."""
    if old_sha and new_sha and old_sha == new_sha:
        return

    old_by_name = {e['name']: e for e in _load_entries(repository, old_sha, tree_cache)}
    new_by_name = {e['name']: e for e in _load_entries(repository, new_sha, tree_cache)}

    for name in set(old_by_name) | set(new_by_name):
        old = old_by_name.get(name)
        new = new_by_name.get(name)
        path = f"{prefix}/{name}" if prefix else name

        if old and new:
            if old['sha'] == new['sha']:
                continue  # identical subtree/blob — prune
            old_is_tree = old['type'] == 'tree'
            new_is_tree = new['type'] == 'tree'
            if old_is_tree and new_is_tree:
                _diff_trees(repository, old['sha'], new['sha'], path, tree_cache, out)
            elif not old_is_tree and not new_is_tree:
                out.append({'path': path, 'status': 'modified',
                            'old_sha': old['sha'], 'new_sha': new['sha']})
            else:  # file <-> directory swap: model as remove + add
                _collect_side(repository, old, path, 'removed', tree_cache, out)
                _collect_side(repository, new, path, 'added', tree_cache, out)
        elif new:
            _collect_side(repository, new, path, 'added', tree_cache, out)
        elif old:
            _collect_side(repository, old, path, 'removed', tree_cache, out)


def _fmt_range(start, stop):
    """Format a unified-diff hunk range (mirrors difflib's own formatter)."""
    length = stop - start
    if length == 0:
        return f"{start},0"
    if length == 1:
        return f"{start + 1}"
    return f"{start + 1},{length}"


def _line_diff(old_text, new_text):
    """Unified line diff -> (lines, additions, deletions)."""
    old_lines = _split_lines(old_text)
    new_lines = _split_lines(new_text)
    matcher = difflib.SequenceMatcher(a=old_lines, b=new_lines, autojunk=False)

    lines = []
    additions = 0
    deletions = 0
    for group in matcher.get_grouped_opcodes(n=3):
        first, last = group[0], group[-1]
        lines.append({
            'kind': 'hunk',
            'text': f"@@ -{_fmt_range(first[1], last[2])} +{_fmt_range(first[3], last[4])} @@",
        })
        for tag, i1, i2, j1, j2 in group:
            if tag == 'equal':
                for off in range(i2 - i1):
                    lines.append({'kind': 'context', 'text': old_lines[i1 + off],
                                  'old': i1 + off + 1, 'new': j1 + off + 1})
            else:
                if tag in ('replace', 'delete'):
                    for i in range(i1, i2):
                        lines.append({'kind': 'del', 'text': old_lines[i], 'old': i + 1})
                        deletions += 1
                if tag in ('replace', 'insert'):
                    for j in range(j1, j2):
                        lines.append({'kind': 'add', 'text': new_lines[j], 'new': j + 1})
                        additions += 1
    return lines, additions, deletions


@extend_schema(
    responses={200: OpenApiTypes.OBJECT, 404: OpenApiTypes.OBJECT},
    summary='Get commit diff',
    description="Return the diff of a commit against its first parent: a list of "
                "added/removed/modified files, each with a unified line diff.",
)
@api_view(['GET'])
@permission_classes([permissions.IsAuthenticated])
def commit_diff(request, owner_id, repo_name, sha):
    """Compute and return the diff for a single commit."""
    repository = get_repository_or_404(owner_id, repo_name, request.user)
    commit = get_object_or_404(Commit, repository=repository, sha=sha)

    parent_sha = commit.parent_shas[0] if commit.parent_shas else None
    old_tree_sha = None
    if parent_sha:
        parent = Commit.objects.filter(repository=repository, sha=parent_sha).first()
        old_tree_sha = parent.tree_sha if parent else None

    tree_cache = {}
    blob_cache = {}
    raw_changes = []
    _diff_trees(repository, old_tree_sha, commit.tree_sha, '', tree_cache, raw_changes)
    raw_changes.sort(key=lambda c: c['path'])

    files = []
    total_additions = 0
    total_deletions = 0
    for change in raw_changes:
        old_text, old_binary = (
            _read_blob(repository, change['old_sha'], blob_cache)
            if change.get('old_sha') else ('', False)
        )
        new_text, new_binary = (
            _read_blob(repository, change['new_sha'], blob_cache)
            if change.get('new_sha') else ('', False)
        )
        binary = old_binary or new_binary

        entry = {
            'path': change['path'],
            'status': change['status'],
            'old_sha': change.get('old_sha'),
            'new_sha': change.get('new_sha'),
            'binary': binary,
            'too_large': False,
            'additions': 0,
            'deletions': 0,
            'lines': [],
        }

        if not binary:
            old_lines = _split_lines(old_text or '')
            new_lines = _split_lines(new_text or '')
            if old_lines and new_lines and len(old_lines) * len(new_lines) > MAX_DIFF_CELLS:
                entry['too_large'] = True
            else:
                lines, additions, deletions = _line_diff(old_text or '', new_text or '')
                entry['lines'] = lines
                entry['additions'] = additions
                entry['deletions'] = deletions
                total_additions += additions
                total_deletions += deletions

        files.append(entry)

    return Response(
        {
            'commit': CommitSerializer(commit).data,
            'parent_sha': parent_sha,
            'files': files,
            'additions': total_additions,
            'deletions': total_deletions,
        },
        status=status.HTTP_200_OK,
    )

File: apps/frontend/src/lib/diff.ts
/**
 * Client-side git diff engine.
 *
 * The Gent API exposes trees and blobs but *no* diff endpoint, so the web app
 * computes "what changed in a commit" itself by comparing the commit's tree
 * against its parent's tree:
 *
 *   1. `diffTrees` walks both trees in lock-step, descending only into subtrees
 *      whose sha differs (identical subtrees are pruned — exactly how git skips
 *      unchanged directories), and emits one {added|removed|modified} record per
 *      changed file.
 *   2. For each changed *text* file we fetch the old/new blob and run `diffText`,
 *      an LCS line diff that produces unified-style hunks with line numbers.
 *
 * Everything here is pure (no network) except via the injected `getTree`
 * accessor, which keeps it easy to cache and test.
 */
import type { Blob, Commit, Tree, TreeEntry } from "@/types/api";

const ZERO_SHA = "0".repeat(64);

/** True for the placeholder sha git/Gent give empty trees & branches. */
export function isEmptyTreeSha(sha: string | null | undefined): boolean {
  return !sha || sha === ZERO_SHA || /^0+$/.test(sha);
}

/* ------------------------------------------------------------------ */
/* Tree-level diff                                                     */
/* ------------------------------------------------------------------ */

export type ChangeStatus = "added" | "removed" | "modified";

export interface RawFileChange {
  path: string;
  status: ChangeStatus;
  oldSha?: string;
  newSha?: string;
}

/** Async tree accessor — typically wraps `gitService.tree` with a cache. */
export type GetTree = (sha: string) => Promise<Tree>;

/** Recursively collect every blob beneath a tree as `path → sha`. */
async function collectBlobs(
  getTree: GetTree,
  sha: string,
  prefix: string,
  out: { path: string; sha: string }[],
): Promise<void> {
  if (isEmptyTreeSha(sha)) return;
  const tree = await getTree(sha);
  for (const e of tree.entries) {
    const p = prefix ? `${prefix}/${e.name}` : e.name;
    if (e.type === "tree") await collectBlobs(getTree, e.sha, p, out);
    else out.push({ path: p, sha: e.sha });
  }
}

/** Emit every blob under one side of a changed entry as added/removed. */
function sideChange(
  path: string,
  status: "added" | "removed",
  sha: string,
): RawFileChange {
  return status === "added"
    ? { path, status, newSha: sha }
    : { path, status, oldSha: sha };
}

async function emitSide(
  getTree: GetTree,
  entry: TreeEntry,
  path: string,
  status: "added" | "removed",
  changes: RawFileChange[],
): Promise<void> {
  if (entry.type === "blob") {
    changes.push(sideChange(path, status, entry.sha));
    return;
  }
  const blobs: { path: string; sha: string }[] = [];
  await collectBlobs(getTree, entry.sha, path, blobs);
  for (const b of blobs) changes.push(sideChange(b.path, status, b.sha));
}

/**
 * Diff two trees into a flat list of file changes. `null` on either side means
 * "this side has no tree" (e.g. the very first commit has no parent), in which
 * case every file on the present side is added/removed.
 */
export async function diffTrees(
  getTree: GetTree,
  oldSha: string | null,
  newSha: string | null,
  prefix = "",
): Promise<RawFileChange[]> {
  if (oldSha && newSha && oldSha === newSha) return [];

  const oldEntries =
    oldSha && !isEmptyTreeSha(oldSha) ? (await getTree(oldSha)).entries : [];
  const newEntries =
    newSha && !isEmptyTreeSha(newSha) ? (await getTree(newSha)).entries : [];

  const oldByName = new Map(oldEntries.map((e) => [e.name, e]));
  const newByName = new Map(newEntries.map((e) => [e.name, e]));
  const names = new Set([...oldByName.keys(), ...newByName.keys()]);

  const changes: RawFileChange[] = [];
  for (const name of names) {
    const o = oldByName.get(name);
    const n = newByName.get(name);
    const p = prefix ? `${prefix}/${name}` : name;

    if (o && n) {
      if (o.sha === n.sha) continue; // identical subtree/blob — prune
      if (o.type === "tree" && n.type === "tree") {
        changes.push(...(await diffTrees(getTree, o.sha, n.sha, p)));
      } else if (o.type === "blob" && n.type === "blob") {
        changes.push({ path: p, status: "modified", oldSha: o.sha, newSha: n.sha });
      } else {
        // file ⇄ directory swap — model as a removal plus an addition
        await emitSide(getTree, o, p, "removed", changes);
        await emitSide(getTree, n, p, "added", changes);
      }
    } else if (n) {
      await emitSide(getTree, n, p, "added", changes);
    } else if (o) {
      await emitSide(getTree, o, p, "removed", changes);
    }
  }
  return changes;
}

/* ------------------------------------------------------------------ */
/* Line-level (text) diff                                              */
/* ------------------------------------------------------------------ */

export type DiffLineKind = "context" | "add" | "del" | "hunk";

export interface DiffLine {
  kind: DiffLineKind;
  text: string;
  oldNumber?: number;
  newNumber?: number;
}

export interface TextDiff {
  lines: DiffLine[];
  additions: number;
  deletions: number;
  /** Set when both sides are large enough that an O(n·m) diff is skipped. */
  tooLarge: boolean;
}

/** ~1.5M DP cells (e.g. 1200×1200) — above this we skip the inline diff. */
const MAX_DIFF_CELLS = 1_500_000;
/** How many unchanged lines to keep around each change. */
const CONTEXT = 3;

function splitLines(text: string): string[] {
  if (text === "") return [];
  return text.replace(/\r\n/g, "\n").replace(/\n$/, "").split("\n");
}

/** Classic LCS backtrack → a full op list (every line tagged). */
function lcsOps(a: string[], b: string[]): DiffLine[] {
  const n = a.length;
  const m = b.length;
  // dp[i][j] = LCS length of a[i:] and b[j:]
  const dp: Uint32Array[] = Array.from({ length: n + 1 }, () => new Uint32Array(m + 1));
  for (let i = n - 1; i >= 0; i--) {
    const row = dp[i];
    const next = dp[i + 1];
    for (let j = m - 1; j >= 0; j--) {
      row[j] = a[i] === b[j] ? next[j + 1] + 1 : Math.max(next[j], row[j + 1]);
    }
  }

  const ops: DiffLine[] = [];
  let i = 0;
  let j = 0;
  let oldNo = 1;
  let newNo = 1;
  while (i < n && j < m) {
    if (a[i] === b[j]) {
      ops.push({ kind: "context", text: a[i], oldNumber: oldNo++, newNumber: newNo++ });
      i++;
      j++;
    } else if (dp[i + 1][j] >= dp[i][j + 1]) {
      ops.push({ kind: "del", text: a[i], oldNumber: oldNo++ });
      i++;
    } else {
      ops.push({ kind: "add", text: b[j], newNumber: newNo++ });
      j++;
    }
  }
  while (i < n) ops.push({ kind: "del", text: a[i++], oldNumber: oldNo++ });
  while (j < m) ops.push({ kind: "add", text: b[j++], newNumber: newNo++ });
  return ops;
}

/** Collapse runs of unchanged context into hunks with `@@ … @@` headers. */
function toHunks(ops: DiffLine[]): DiffLine[] {
  const changed = ops
    .map((o, i) => (o.kind !== "context" ? i : -1))
    .filter((i) => i >= 0);
  if (changed.length === 0) return [];

  const windows: [number, number][] = [];
  for (const idx of changed) {
    const start = Math.max(0, idx - CONTEXT);
    const end = Math.min(ops.length - 1, idx + CONTEXT);
    const last = windows[windows.length - 1];
    if (last && start <= last[1] + 1) last[1] = Math.max(last[1], end);
    else windows.push([start, end]);
  }

  const out: DiffLine[] = [];
  for (const [s, e] of windows) {
    const slice = ops.slice(s, e + 1);
    const firstOld = slice.find((o) => o.oldNumber != null)?.oldNumber ?? 0;
    const firstNew = slice.find((o) => o.newNumber != null)?.newNumber ?? 0;
    const oldCount = slice.filter((o) => o.oldNumber != null).length;
    const newCount = slice.filter((o) => o.newNumber != null).length;
    out.push({
      kind: "hunk",
      text: `@@ -${firstOld},${oldCount} +${firstNew},${newCount} @@`,
    });
    out.push(...slice);
  }
  return out;
}

/** Unified line diff between two text blobs. */
export function diffText(oldText: string, newText: string): TextDiff {
  const a = splitLines(oldText);
  const b = splitLines(newText);

  // Pure add / pure remove are cheap (one side empty) and never hit the cap.
  if (a.length > 0 && b.length > 0 && a.length * b.length > MAX_DIFF_CELLS) {
    return { lines: [], additions: 0, deletions: 0, tooLarge: true };
  }

  const ops = lcsOps(a, b);
  const additions = ops.reduce((s, o) => s + (o.kind === "add" ? 1 : 0), 0);
  const deletions = ops.reduce((s, o) => s + (o.kind === "del" ? 1 : 0), 0);
  return { lines: toHunks(ops), additions, deletions, tooLarge: false };
}

/* ------------------------------------------------------------------ */
/* Blob decoding + assembled per-commit diff types                     */
/* ------------------------------------------------------------------ */

/** Decode a blob for diffing. base64 blobs are treated as binary (not shown). */
export function blobToText(blob: Blob | undefined): {
  text: string | null;
  binary: boolean;
} {
  if (!blob) return { text: null, binary: false };
  if (blob.encoding === "utf-8") return { text: blob.content ?? "", binary: false };
  return { text: null, binary: true };
}

export interface FileDiff extends RawFileChange {
  binary: boolean;
  tooLarge: boolean;
  additions: number;
  deletions: number;
  lines: DiffLine[];
}

export interface CommitDiffResult {
  commit: Commit;
  parentSha: string | null;
  files: FileDiff[];
  additions: number;
  deletions: number;
}
